### PR TITLE
Add caching for the rest of the islamic calendars (‫عید مبارک!!‬)

### PR DESCRIPTION
**Pull request by Manishearth**
Fixes https://github.com/unicode-org/icu4x/issues/3933

Uses work from https://github.com/unicode-org/icu4x/pull/4770 on the remaining 3 calendars

عید مبارک!!

### CODE REVIEW

File: components/calendar/src/islamic.rs [hunk 1]
```diff
@@ -149,32 +156,122 @@ impl IslamicObservational {
 }
 
 impl IslamicCivil {
-    /// Construct a new [`IslamicCivil`] without any precomputed calendrical calculations.
+    /// Creates a new [`IslamicCivil`] with some precomputed calendrical calculations.
```
Comment (sffc) [inline components/calendar/src/islamic.rs:159]: here and elsewhere: `with compiled data`

File: components/calendar/src/islamic.rs [hunk 2]
```diff
@@ -741,81 +881,35 @@ impl<A: AsCalendar<Calendar = IslamicUmmAlQura>> DateTime<A> {
...
+    fn days_in_provided_year(_year: i32, year_info: IslamicYearInfo) -> u16 {
+        year_info.packed_data.days_in_year()
     }
 
-    fn is_leap_year(year: i32, _data: ()) -> bool {
-        (14 + 11 * year).rem_euclid(30) < 11
+    // As an true lunar calendar, it does not have leap years.
+    fn is_leap_year(_year: i32, _year_info: IslamicYearInfo) -> bool {
+        false
     }
```
Comment (sffc) [inline components/calendar/src/islamic.rs:907]: I'm confused. Why is the logic changing for is_leap_year?

File: components/calendar/src/islamic.rs [hunk 3]
```diff
@@ -741,81 +881,35 @@ impl<A: AsCalendar<Calendar = IslamicUmmAlQura>> DateTime<A> {
...
-            2 | 4 | 6 | 8 | 10 => 29,
-            12 if Self::is_leap_year(year, ()) => 30,
-            12 => 29,
-            _ => 0,
-        }
+    type YearInfo = IslamicYearInfo;
+
+    fn month_days(_year: i32, month: u8, year_info: IslamicYearInfo) -> u8 {
+        year_info.packed_data.days_in_month(month)
     }
```
Comment (sffc) [inline components/calendar/src/islamic.rs:894]: This calendar seems to have a nice formula for month lengths. Why make that part data-driven? I thought we would only make expensive computations be data-driven.

File: components/calendar/src/islamic.rs [hunk 4]
```diff
@@ -1030,40 +1130,29 @@ impl<A: AsCalendar<Calendar = IslamicCivil>> DateTime<A> {
...
-            1 | 3 | 5 | 7 | 9 | 11 => 30,
-            2 | 4 | 6 | 8 | 10 => 29,
-            12 if Self::is_leap_year(year, ()) => 30,
-            12 => 29,
-            _ => 0,
-        }
+    type YearInfo = IslamicYearInfo;
+
+    fn month_days(_year: i32, month: u8, year_info: IslamicYearInfo) -> u8 {
+        year_info.packed_data.days_in_month(month)
```
Comment (sffc) [inline components/calendar/src/islamic.rs:1136]: Same in this class: why are you changing a straightforward match into a data-driven function?

File: components/calendar/src/islamic.rs [hunk 5]
```diff
@@ -1030,40 +1130,29 @@ impl<A: AsCalendar<Calendar = IslamicCivil>> DateTime<A> {
...
-        }
+    fn days_in_provided_year(_year: i32, year_info: IslamicYearInfo) -> u16 {
+        year_info.packed_data.days_in_year()
     }
 
-    fn is_leap_year(year: i32, _data: ()) -> bool {
-        (14 + 11 * year).rem_euclid(30) < 11
+    // As an true lunar calendar, it does not have leap years.
+    fn is_leap_year(_year: i32, _year_info: IslamicYearInfo) -> bool {
+        false
```
Comment (sffc) [inline components/calendar/src/islamic.rs:1149]: Same here: why is `is_leap_year` changing?

**Review by sffc — COMMENTED** [review]
.

File: components/calendar/src/islamic.rs (hunk 3 above)
Comment (Manishearth) [reply components/calendar/src/islamic.rs:894]: Ah, now I remember. We only wanted to optimize UAQ and observational.

File: components/calendar/src/islamic.rs (hunk 2 above)
Comment (Manishearth) [reply components/calendar/src/islamic.rs:907]: I'm not actually sure what we should do here; on one hand it's a lunar calendar without leap months, on the other hand you _can_ have a notion of regular/leap years here, and for the other calendars.

I think I'll have all Islamic calendars do the 354/355 leap year thing for now.

**Review by sffc — APPROVED** [review]
Much improved; yeah, let's only use the precompiled data solution when we have astonomy going on. The formulas implementing `islamic-tbla` and `islamic-civil` are very straightforward.

Thought: I wonder if we could make `islamic-umalqura` and `islamic` fall back to one of the formulaic calendars for distant dates instead of carrying around the astronomy code at runtime.

File: components/calendar/src/islamic.rs [hunk 6]
```diff
@@ -76,8 +79,7 @@ pub struct IslamicObservational {
 ///
 /// This calendar is a pure lunar calendar with no leap months. It uses month codes
 /// `"M01" - "M12"`.
-#[derive(Clone, Debug, Hash, Eq, PartialEq, PartialOrd, Ord)]
-#[non_exhaustive] // we'll be adding precompiled data to this
+#[derive(Clone, Debug, Default)]
```
[the remark this answers is not in the document]
Comment (Manishearth) [reply components/calendar/src/islamic.rs:82]: We shouldn't have them in the first place, it's a mistake.

[the remark this answers is not in the document]
Comment (Manishearth) [reply components/calendar/src/islamic.rs:82]: I'll merge this for now since it has datagen but we can revert this change if we want.

**Comment by sffc** [comment]
```
error: use of deprecated associated function `icu_calendar::islamic::IslamicCivil::new_always_calculating`: Precomputation not needed for this calendar
   --> components/calendar/benches/datetime.rs:189:47
    |
189 |         icu::calendar::islamic::IslamicCivil::new_always_calculating(),
    |                                               ^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `-D deprecated` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(deprecated)]`

error: use of deprecated associated function `icu_calendar::islamic::IslamicCivil::new_always_calculating`: Precomputation not needed for this calendar
   --> components/calendar/benches/datetime.rs:198:55
    |
198 |                 icu::calendar::islamic::IslamicCivil::new_always_calculating(),
    |                                                       ^^^^^^^^^^^^^^^^^^^^^^

error: use of deprecated associated function `icu_calendar::islamic::IslamicTabular::new_always_calculating`: Precomputation not needed for this calendar
   --> components/calendar/benches/datetime.rs:209:49
    |
209 |         icu::calendar::islamic::IslamicTabular::new_always_calculating(),
    |                                                 ^^^^^^^^^^^^^^^^^^^^^^

error: use of deprecated associated function `icu_calendar::islamic::IslamicTabular::new_always_calculating`: Precomputation not needed for this calendar
   --> components/calendar/benches/datetime.rs:218:57
    |
218 |                 icu::calendar::islamic::IslamicTabular::new_always_calculating(),
    |                                                         ^^^^^^^^^^^^^^^^^^^^^^

error: could not compile `icu_calendar` (bench "datetime") due to 4 previous erro
```

**Review by sffc — APPROVED** [review]

**Pull request merged**